File: 2975-maximum-square-area-by-removing-fences-from-a-field/2975-maximum-square-area-by-removing-fences-from-a-field.py
class Solution:
    def maximizeSquareArea(self, m: int, n: int, hFences: List[int], vFences: List[int]) -> int:
        hFences, vFences = [1] + hFences + [m], [1] + vFences + [n]
        
        hFences.sort()
        vFences.sort()

        hSet = set()
        for i in range(0, len(hFences)-1):
            for j in range(i + 1, len(hFences)):
                s = hFences[j] - hFences[i]
                hSet.add(s)
        
        max_s = -1
        for i in range(0, len(vFences)-1):
            for j in range(i + 1, len(vFences)):
                s = vFences[j] - vFences[i]
                if s in hSet:
                    max_s = max(max_s, s)

        if max_s == -1:
            return -1
        
        MOD = 1_000_000_007

        return (max_s * max_s) % MOD

# array, hash map
# time: O(h^2 + v^2)
# space: O(h^2 + h + v)

File: 2975-maximum-square-area-by-removing-fences-from-a-field/2975-maximum-square-area-by-removing-fences-from-a-field.go
func maximizeSquareArea(m int, n int, hFences []int, vFences []int) int {
    // add unremovable fence (boundary) for hFences and vFences
    hFences = append([]int{1}, append(hFences, m)...)
    vFences = append([]int{1}, append(vFences, n)...)

    sort.Slice(hFences, func(i, j int) bool { return hFences[i] < hFences[j] })
    sort.Slice(vFences, func(i, j int) bool { return vFences[i] < vFences[j] })

    hSet := map[int]struct{}{}
    for i := 0; i < len(hFences)-1; i++ {
        for j := i + 1; j < len(hFences); j++ {
            s := hFences[j] - hFences[i]
            if _, exists := hSet[s]; !exists {
                hSet[s] = struct{}{}
            }
        }
    }

    maxS := -1 // max side
    for i := 0; i < len(vFences)-1; i++ {
        for j := i + 1; j < len(vFences); j++ {
            s := vFences[j] - vFences[i]
            if _, exists := hSet[s]; exists {
                maxS = max(maxS, s)
            }
        }
    }

    if maxS == -1 {
        return -1
    }

    const MOD = 1_000_000_007

    return (maxS * maxS) % MOD
}

// array, hash map
// time: O(h^2 + v^2)
// space: O(h^2 + h + v)